### PR TITLE
Issues/91 photo detail preview

### PR DESCRIPTION
**Pull request by aerych**
Refs #91 

In this patch we're adding the ability to view/preview a Photo asset, and to edit its caption and alt text.  

This patch does the following: 
- Introduces a new PhotoDetailViewController for previewing a Photo, including a caption and alt text. This is inspired by the MediaItemViewController in WPiOS.
- Introduces a new TextFieldViewController for editing text field. This is inspired by the SettingsTextViewController in WPiOS. 
- Introduces a new ImageViewController) for viewing and zooming an image. This is modeled after the WPImageViewController in WPiOS.
- Removes the AttachmentInfo model and moves its fields into the StoryAsset model. This entity was ultimately a poor fit.
- Switches the cells used on the FolderViewController to the common TextFieldTableViewCell. (One text field cell to rule them all!)

It has ended up being a larger patch than I expected.  A good chunk is boiler plate for core data, storyboard and xibs, but even so there is more to review than I'd prefer so let's try splitting the code review between a couple of folks this time. 

@jleandroperez would you mind taking a peek at and testing PhotoDetailViewController, TextFieldViewController and their related changes.  @jkmassel would you mind taking a peek at and testing the WPImageViewController change?

### To test: 
Prep:  
1. This makes some core data changes so start from a fresh install (no migrations while in dev).  
2. Log in with a wpcom site.
3. Open the Photos app and use the share feature to copy a few photos.
4. Open the Files app, browse to the Newspack app's folder, then the folder for the site you added. Find the story folder inside the site and paste the photos inside.

### PhotoDetails: 
- [x] From the Assets screen, confirm that tapping on a Photo asset shows the PhotoDetailViewController for the correct Photo.
- [x] On the PhotoDetailScreen tap on the caption field. Confirm you're taken to a TextFieldViewController.
- [x] Confirm the Save button on the TextFieldViewController is disabled initially and is enabled when the caption text is changed.
- [x] Enter a caption and save.  Confirm that the caption appears on the details screen,
- [x] Confirm the caption appears in the Photo cell on the Assets screen. (The layout is a little tight. This will get sorted later.)
- [x] Repeat these steps to confirm you can edit the caption and the edit appears as expected.
- [x] Edit a caption but this time use the cancel button. Confirm the change is not made. 
- [x] Repeat these steps for the photo's alt text.  Confirm every thing works as expected. (Note that alt text is not shown on  the asset cells.)

### Story Folder Regression Check
- [x] On the Story folders screen, tap to create a new story folder (or swipe to edit an existing folder. Confirm the add/edit screen appears, that its cell looks correct, and there is no crash or warning. 

### ImageViewController
- [ ] From the Photo detail screen tap on the image.  Confirm you are presented with a larger version of the image in a "lightbox" sort of view. 
- [ ] Confirm you can double tap to zoom in and out.
- [ ] Confirm you can pinch to zoom.
- [ ] Confirm you can tap once to dismiss
- [ ] Confirm you can swipe / fling to dismiss. 
- [ ] Confirm the image looks good whether its a portrait or landscape image. 
- [ ] Confirm the image looks good when rotating the device. 

### Examples:
Simulator Screen Shot - iPhone 11 Pro Max - 2020-08-26 at 15 29 48
Simulator Screen Shot - iPhone 11 Pro Max - 2020-08-26 at 15 29 58
Simulator Screen Shot - iPhone 11 Pro Max - 2020-08-26 at 15 30 02
Simulator Screen Shot - iPhone 11 Pro Max - 2020-08-26 at 15 30 04
Simulator Screen Shot - iPhone 11 Pro Max - 2020-08-26 at 15 30 11

### CODE REVIEW

**Review by jkmassel — APPROVED** [review]
> Open the Files app, browse to the Newspack app's folder, then the folder for the site you added. Find the story folder inside the site and paste the photos inside.

I tried this on my carry phone, and after login I didn't have a Newspack app folder in the Files app (even after adding a photo using the toolbar icon). That didn't affect my ability to test, but I figured it might be worth pointing out.

✅ From the Photo detail screen tap on the image. Confirm you are presented with a larger version of the image in a "lightbox" sort of view.
✅ Confirm you can double tap to zoom in and out. [zooming in went _much_ farther in than I expected, but that's probably outside the scope of this PR]
✅ Confirm you can pinch to zoom.
✅ Confirm you can tap once to dismiss
✅ Confirm you can swipe / fling to dismiss.
✅ Confirm the image looks good whether its a portrait or landscape image.
✅ Confirm the image looks good when rotating the device.

I looked at the code for `ImageViewController` and it looks very straightforward, no concerns there!

A few nits which may or may not be part of this PR, but I figured I'd note them because I noticed them:
- Performing a "fling" action didn't always track my finger correctly as I performed it (I think this is the dependency, not this code)
- The cell for the image briefly glows blue when tapped to open the `ImageViewController`
- There's a bit of a delay when opening and flinging-to-dismiss, but not when tapping to dismiss.

File: Newspack/Newspack/Controllers/ImageViewController.swift [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+/// or zoomed in/out by double tapping.
+/// Tap once to dismiss.
+///
+class ImageViewController: UIViewController {
+
+    @IBOutlet var scrollView: UIScrollView!
+    @IBOutlet var imageView: UIImageView!
+
+    lazy var flingableViewHandler: FlingableViewHandler = {
+        let handler = FlingableViewHandler(targetView: self.scrollView)
```
Comment (jleandroperez) [inline Newspack/Newspack/Controllers/ImageViewController.swift:14]: I LOVE the `FlingableViewHandler` behavior!!!

Comment (jleandroperez) [reply Newspack/Newspack/Controllers/ImageViewController.swift:14]: BTW maybe the `lazy var`(s) should be private?

File: Newspack/Newspack/Controllers/ImageViewController.swift [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+    func viewForZooming(in scrollView: UIScrollView) -> UIView? {
+        return imageView
+    }
+
+    func scrollViewDidZoom(_ scrollView: UIScrollView) {
+        let w = scrollView.bounds.size.width - scrollView.contentSize.width
+        let h = scrollView.bounds.size.height - scrollView.contentSize.height
+        let x = max(w / 2, 0)
+        var y = max(h / 2, 0)
```
Comment (jleandroperez) [inline Newspack/Newspack/Controllers/ImageViewController.swift]: Nipticky: Maybe `round` on the x/y, just in case?

Plus: `y` could also be `let` perhaps?

**Review by jleandroperez — APPROVED** [review]
@aerych Looks BEAUTIFUL!!

✅ Unit Tests look great
✅ No regressions in the Folder Handling mechanism(s)
✅ Photo Details: Everything is green!

Since I was in the neighborhood, I take a look at the ImageViewController tests as well. Few notes:

1. The Image Row gets highlighted. Noticeable if the image doesn't match the row's width:


2. Double Press to Zoom In applies a bit of a super extreme zoom level (magnifies ... perhaps, 5 points square, in regular zoom, to fullscreen)

3. Double Press to Zoom: Once Zoomed In, you cannot move around the picture (dragging causes the Flingable behavior to kick in). **Same note for Pinch to Zoom!**

4. When you start swiping to dismiss, if you decide to cancel the action, the image is centered back, immediately, with no animation. 


Usability wise, it's solid, and looks great. Good to address the notes up there in a follow up PR, up to you sir!!

:shipit:

File: Newspack/Newspack/Controllers/ImageViewController.swift (hunk 2 above)
Comment (aerych) [reply Newspack/Newspack/Controllers/ImageViewController.swift]: Arg! I made this a var while wrestling with a positioning bug due to the status bar/nav bar being included in the initial layout pass.  I solved that issue a different way and forgot to change it back.  Well spotted.

**Comment by aerych** [comment]
> I tried this on my carry phone, and after login I didn't have a Newspack app folder in the Files app 

In the Files app, maybe check if the Browse tab is selected, then tap back until you reach the main Browse screen. Then tap on On My iPhone. There _should_ be a Newspack folder there. The app would probably go 💥 if it couldn't make it. 

> Performing a "fling" action didn't always track my finger correctly as I performed it (I think this is the dependency, not this code)

I've noticed this. It is all part of a dependency so I haven't looked closely at it.  I might switch the behavior to a simple swipe down to delete if its feeling too wonky. Good to flag it now so I can keep it in mind.
> The cell for the image briefly glows blue when tapped to open the ImageViewController

Ah! Yes. That is the current cell selection behavior and is expected.
> There's a bit of a delay when opening and flinging-to-dismiss, but not when tapping to dismiss

Well spotted.  The opening delay (tho not ideal) is expected and a consequence of they way I've solved an issue where the image's starting position and size was off due to the first layout pass happening before being added to the view hierarchy, ergo not having the correct geometry. I'm tolerating it for now but would love to revisit later.  

Looks like the fling to dismiss can be sped up a bit by removing the dispatch call.  I saw this was how it was implemented in WPiOS and just ran with it but in testing now it looks like it's not needed at all.

Thanks @jkmassel!

**Comment by aerych** [comment]
> The Image Row gets highlighted. Noticeable if the image doesn't match the row's width:

You and @jkmassel both have flagged this so I'm rethinking letting the selection behavior show.  We'll disable it for this cell. 
> Double Press to Zoom In applies a bit of a super extreme zoom level (magnifies ... perhaps, 5 points square, in regular zoom, to fullscreen)

This is consistent with the behavior in WPiOS so I think we can leave it for now. 
> Double Press to Zoom: Once Zoomed In, you cannot move around the picture (dragging causes the Flingable behavior to kick in). Same note for Pinch to Zoom!

doH!  My bad. I completely forgot to add the call to disable the flinging behavior when zoomed.  Fixing!

> When you start swiping to dismiss, if you decide to cancel the action, the image is centered back, immediately, with no animation.

I think the fix for this would be to patch  the control in the WordPressUI pod, or drop the flinging action in favor of a simple swipe to dismiss like you see with modal views. We can revisit in a future design pass. 

Thank you @jleandroperez !

**Comment by jkmassel** [comment]
> Then tap on On My iPhone

Ah, yep! This was the part I missed – I thought it'd be in iCloud. It's there!

**Comment by aerych** [comment]
> I thought it'd be in iCloud

Future feature! :D

**Comment by aerych** [comment]
Thanks again @jleandroperez and @jkmassel !! 🙇‍♂️

**Pull request merged**